src/cluster/placement/algo/mirrored_test.go

```diff
@@ -646,37 +429,318 @@ func TestMirrorRemoveAndRevertAfterCutover(t *testing.T) {
+
+	// Third replace: i1 replaced by i6, is overlapping with first replace
+	// because i1 has a pending replacement peer i5 with initializing shards that are not cutover.
+	p2, err = a.ReplaceInstances(p1, []string{"i1"}, []placement.Instance{i6})
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "Not cutover")
+	assert.Nil(t, p2)
+
+	assert.Equal(t, uint32(2), p1.MaxShardSetID())
+	assert.True(t, allLeaving(p1, []placement.Instance{i1}, nowNanos))
+	assert.True(t, allInitializing(p1, []string{"i5"}, nowNanos))
+
+	// Fourth replace: i3 replaced by i6, is not overlapping with first replace and must go through.
+	i6.SetShardSetID(i3.ShardSetID())
+	p2, err = a.ReplaceInstances(p1, []string{"i3"}, []placement.Instance{i6})
+	require.NoError(t, err)
+	require.NoError(t, placement.Validate(p2))
+	assert.Equal(t, uint32(2), p2.MaxShardSetID())
+	assert.True(t, allLeaving(p2, []placement.Instance{i1, i3}, nowNanos))
+	assert.True(t, allInitializing(p2, []string{"i5", "i6"}, nowNanos))
+}
+
+func TestMirrorRevertOfReplaceMustMatch(t *testing.T) {
+	i1 := newTestInstance("i1").SetShardSetID(1)
+	i2 := newTestInstance("i2").SetShardSetID(1)
+	i3 := newTestInstance("i3").SetShardSetID(2)
+	i4 := newTestInstance("i4").SetShardSetID(2)
+	instances := []placement.Instance{i1, i2, i3, i4}
+
+	numShards := 100
+	ids := make([]uint32, numShards)
+	for i := 0; i < len(ids); i++ {
+		ids[i] = uint32(i)
+	}
+
+	now := time.Now()
+	nowNanos := now.UnixNano()
+	shardCutoverTime := now.Add(time.Hour).UnixNano()
+	a := NewAlgorithm(placement.NewOptions().
+		SetIsMirrored(true).
+		SetShardCutoverNanosFn(func() int64 { return shardCutoverTime }).
+		SetShardCutoffNanosFn(func() int64 { return shardCutoverTime }))
+	p, err := a.InitialPlacement(instances, ids, 2)
+	require.NoError(t, err)
+	require.NoError(t, placement.Validate(p))
+	assert.Equal(t, uint32(2), p.MaxShardSetID())
+
+	p, _, err = a.MarkAllShardsAvailable(p)
+	assert.NoError(t, err)
+
+	// First replace: i1 replaced by i1a.
+	i1a := newTestInstance("i1a").
+		SetShardSetID(i1.ShardSetID()).
+		SetWeight(i1.Weight())
+	p1, err := a.ReplaceInstances(p, []string{"i1"}, []placement.Instance{i1a})
+	require.NoError(t, err)
+	require.NoError(t, placement.Validate(p1))
+	assert.Equal(t, uint32(2), p1.MaxShardSetID())
+	assert.True(t, allLeaving(p1, []placement.Instance{i1}, nowNanos))
+	assert.True(t, allInitializing(p1, []string{"i1a"}, nowNanos))
+
+	// Second replace: i3 replaced by i3a.
+	i3a := newTestInstance("i3a").
+		SetShardSetID(i3.ShardSetID()).
+		SetWeight(i3.Weight())
+	p2, err := a.ReplaceInstances(p1, []string{"i3"}, []placement.Instance{i3a})
+	require.NoError(t, err)
+	require.NoError(t, placement.Validate(p2))
+	assert.Equal(t, uint32(2), p2.MaxShardSetID())
+	assert.True(t, allLeaving(p2, []placement.Instance{i1, i3}, nowNanos))
+	assert.True(t, allInitializing(p2, []string{"i1a", "i3a"}, nowNanos))
+
+	// Revert of non-matching replace pairs must fail.
```

Comment:
These look like separate test cases to me--am I right about that? If so, might be worth doing them as `t.Run()`'s? 

Couple advantages:

- Individually addressable
- Less possibility of weird interactions between tests

Reply:
Done, thanks.